### ISSUE TITLE
Show kernel type in list of notebooks

### ISSUE THREAD

**Issue by payne92** [ipython/ipython#7831]
UI enhancement:  consider showing the kernel type (Python, Lua, bash, etc.) in the main notebook page (the list of notebooks).

**Comment by takluyver** [comment]
Unfortunately, that would involve reading every notebook in the directory,
which would show down showing the dashboard.

**Comment by payne92** [comment]
Agreed -- it probably needs a cache to speed loading.

With the generalization of kernels (and proliferation of different types of kernels), I'm guessing you'll be getting feature pressure for things like this (various notebook attributes presented in the main notebook page).

**Comment by takluyver** [comment]
Yeah, this isn't the first time this has been requested, and I'm sure it won't be the last. :-)

Designing a mechanism to do this with a cache isn't something that sounds appealing, given our other priorities. Some of the things we'd need to think about:

- What do we do when browsing to a directory where that information isn't cached? Hold up the listing until we've read all the notebooks, or have code to render it with that information missing?
- If we render with some information missing, do we update the view as the information is read, or only when the listing is next loaded?
- When do we invalidate the cache? ("There are only two hard problems in computer science: naming things, cache invalidation, and off by one errors.")
- Entering a directory with 10k notebooks shouldn't lock up the server while it reads every single one, so we need to fill the cache asynchronously. Do we dance around the server event loop, or run a separate thread or a separate process? If it's a separate thread/process, do we start a new one for each directory, or keep one around and send it messages to index directories?

I can hazard a guess at what the answers would be, but I suspect this is probably something we won't get to until after 4.0.
